File: HBPhoneNumberFormatter/Classes/HBPhoneNumberFormatter.m
//
//  HBNumberFormatting.m
//  HBNumberFormatting
//
//  Created by Hayk Brsoyan on 6/18/17.
//  Copyright © 2017 Hayk Brsoyan. All rights reserved.
//

#import "HBPhoneNumberFormatter.h"
#import "HBCountryCode.h"

@interface HBPhoneNumberFormatter ()
@property (nonatomic) NSDictionary *symbols;
@property (nonatomic) int numberCount;
@property (nonatomic) BOOL isDelete;
@property (nonatomic) NSString *prefix;
@end

@implementation HBPhoneNumberFormatter

- (instancetype)initWithFormatting:(NSString *)formatting {
    if(!(self = [super init])) {
        return nil;
    }
    
    if (!formatting) {
        NSLog(@"HB Please set up formatting");
        return nil;
    }
    
    // Default shake config
    self.isShake = YES;
    self.shakeDuration = 0.1;
    self.shakeRepeatCount = 2;
    self.shakeSize = 5;
    
    NSUInteger len = [formatting length];
    unichar buffer[len + 1];
    
    [formatting getCharacters:buffer range:NSMakeRange(0, len)];
    
    NSMutableDictionary *symbols = [[NSMutableDictionary alloc] init];
    for(int i = 0; i < len; i++) {
        NSString *character = [NSString stringWithFormat:@"%C", buffer[i]];
        if (![self isNumber:character]) {
            symbols[[@(i) stringValue]] = character;
        }
    }
    self.symbols = [symbols copy];
    self.numberCount = (int)([formatting length] - self.symbols.count);
    
    return self;
}

- (void)setCountryName:(NSString *)name textField:(UITextField *)textField {
    NSString *code = [HBCountryCode countryCodeWithName:name];
    if (!code) {
        self.prefix = [HBCountryCode countryCodeWithName:@"United States"];
        NSLog(@"We dosn't finde %@ .HB Default prefix is United States",name);
    } else {
        self.prefix = code;
    }
    textField.text = [NSString stringWithFormat:@"%@",code];
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {
    // typing symbol is number
    if (![self isNumber:string]) {
        [self shakeTextField:textField];
        return NO;
    }
    if (range.length > 0 && [self.prefix isEqualToString:textField.text]) {
        [self shakeTextField:textField];
        return NO;
    }
    if(range.length > 0) {
        self.isDelete = YES;
    } else {
        self.isDelete = NO;
    }
    
    NSString *number = [self getNumber:textField.text];
    NSUInteger numberLength = number.length;
    if(numberLength == self.numberCount) {
        NSString *symbol = [self symbolForIndex:textField.text.length];
        if (symbol) {
            textField.text = [NSString stringWithFormat:@"%@%@", textField.text, symbol];
            symbol = [self symbolForIndex:textField.text.length];
            [self addSymbolInTextField:textField symbol:symbol];
        }
        
        if (range.location == self.numberCount + self.symbols.count + self.prefix.length) {
            [self shakeTextField:textField];
        }
        if(range.length == 0) {
            return NO;
        }
    }
    
    // TextField is Empty
    if (textField.text.length == self.prefix.length) {
        if (self.numberCount + self.symbols.count > textField.text.length) {
            NSString *symbol = [self symbolForIndex:self.prefix.length];
            if (symbol) {
                if (textField.text.length != 0) {
                    textField.text = [NSString stringWithFormat:@"%@%@",textField.text ,symbol];
                } else {
                    textField.text = [NSString stringWithFormat:@"%@" ,symbol];
                }
                symbol = [self symbolForIndex:textField.text.length];
                [self addSymbolInTextField:textField symbol:symbol];
                
                textField.text = [NSString stringWithFormat:@"%@%@", textField.text, string];
                symbol = [self symbolForIndex:textField.text.length];
                
                [self addSymbolInTextField:textField symbol:symbol];
                
                return NO;
            }
        }
        return NO;
    }
    
    // When added symbol in textField
    if (textField.text.length == range.location) {
        if (self.numberCount + self.symbols.count + self.prefix.length > textField.text.length) {
            NSString *symbol = [self symbolForIndex:textField.text.length];
            if (symbol) {
                textField.text = [NSString stringWithFormat:@"%@%@", textField.text, symbol];
                symbol = [self symbolForIndex:textField.text.length];
                [self addSymbolInTextField:textField symbol:symbol];
            }
            if (!self.isDelete && textField.text.length > 0) {
                textField.text = [NSString stringWithFormat:@"%@%@",textField.text,string];
            }
            symbol = [self symbolForIndex:textField.text.length];
            if (symbol) {
                textField.text = [NSString stringWithFormat:@"%@%@", textField.text, symbol];
                symbol = [self symbolForIndex:textField.text.length];
                [self addSymbolInTextField:textField symbol:symbol];
            }

            if (self.isDelete) {
                textField.text = [NSString stringWithFormat:@"%@%@",textField.text,string];
            }
            return NO;
        }
        return NO;
    }
    
    // Delete text from textField
    if(range.length > 0) {
        if (range.location >= self.prefix.length) {
            if (textField.text.length == range.location) {
                textField.text = [textField.text substringToIndex:textField.text.length - 1];
            } else {
                textField.text = [textField.text substringToIndex:range.location];
            }
            NSString *symbol = [self symbolForIndex:textField.text.length - 1];
            if (symbol) {
                do {
                    textField.text = [textField.text substringToIndex:(textField.text.length - 1)];
                    symbol = [self symbolForIndex:(textField.text.length - 1)];
                } while(symbol);
            }
        } else {
            textField.text = [textField.text substringToIndex:self.prefix.length];
        }

        return NO;
    }
    return YES;
}

- (void)addSymbolInTextField:(UITextField *)textField symbol:(NSString *)symbol {
    while (symbol) {
        textField.text = [NSString stringWithFormat:@"%@%@", textField.text, symbol];
        symbol = [self symbolForIndex:textField.text.length];
    }
}

- (NSString *)symbolForIndex:(NSUInteger)index {
    index = index - self.prefix.length;
    for (NSString *key in self.symbols.allKeys) {
        if ([key isEqualToString:[@(index) stringValue]]) {
            return self.symbols[key];
        }
    }
    return nil;
}

- (NSString *)getNumber:(NSString *)mobileNumber {
    NSUInteger len = [mobileNumber length];
    unichar buffer[len + 1];
    [mobileNumber getCharacters:buffer range:NSMakeRange(0, len)];
    NSString *number;
    for(NSInteger i = self.prefix.length; i < len; i++) {
        NSString *character = [NSString stringWithFormat:@"%C", buffer[i]];
        if ([self isNumber:character]) {
            if (number.length != 0) {
                number = [NSString stringWithFormat:@"%@%@",number,character];
            } else {
                number = [NSString stringWithFormat:@"%@",character];
            }
        }
    }
    return number;
}

- (BOOL)isNumber:(NSString *)number {
    NSCharacterSet *digits = [[NSCharacterSet characterSetWithCharactersInString:@"0123456789"] invertedSet];
    return [number rangeOfCharacterFromSet:digits].location == NSNotFound;
}

- (BOOL)numberIsValidPhoneText:(NSString *)phoneText {
    NSInteger length = [self getNumber:phoneText].length;
    NSString *text = [self getNumber:phoneText];
    if (length == self.numberCount && [self isNumber:text]) {
        return YES;
    } else {
        return NO;
    }
}

- (void)shakeTextField:(UITextField *)textField {
    if (self.isShake) {
        CABasicAnimation *shake = [CABasicAnimation animationWithKeyPath:@"position"];
        [shake setDuration:self.shakeDuration];
        [shake setRepeatCount:self.shakeRepeatCount];
        [shake setAutoreverses:YES];
        [shake setFromValue:[NSValue valueWithCGPoint:
                             CGPointMake(textField.center.x - self.shakeSize,textField.center.y)]];
        [shake setToValue:[NSValue valueWithCGPoint:
                           CGPointMake(textField.center.x + self.shakeSize, textField.center.y)]];
        [textField.layer addAnimation:shake forKey:@"position"];
    }
}

@end

File: HBPhoneNumberFormatter-Swift/HBPhoneNumberFormatter.swift
//
//  HBPhoneNumberFormatter.swift
//  HBPhoneNumberFormatter
//
//  Created by Hovhannes Stepanyan on 6/11/18.
//

import UIKit

class HBPhoneNumberFormatter: NSObject {
    var isShake: Bool = true
    var shakeSize: CGFloat = 5.0
    var shakeDuration: CGFloat = 0.1
    var shakeRepeatCount: CGFloat = 2.0
    
    /// Create formatter and setup your custom formatting, like this @"(111) 1233-222"
    init(formatting: String) {
        super.init()
        let count = formatting.count
        var buffer = [unichar]()
        let _formatting = NSString(string: formatting)
        _formatting.getCharacters(&buffer, range: NSMakeRange(0, count))
        for char in buffer {
            let character = String(char)
            if !self.isNumber(character) {
                symbols[String(buffer.index(of: char)!)] = character
            }
        }
        self.numberCount = formatting.count - symbols.count
    }
    
    /// Call this method from your uiviewcontroller <UITextFieldDelegate> similar method.
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        if !self.isNumber(string) {
            self.shake(textField: textField)
            return false
        }
        if range.length > 0 && self.prefix == textField.text {
            self.shake(textField: textField)
            return false
        }
        if range.length > 0 {
            self.isDelete = true
        } else {
            self.isDelete = false
        }
        let number = self.getNumber(textField.text)
        let numberCount = number.count
        if numberCount == self.numberCount {
            if var symbol = self.symbolFor(textField.text?.count) {
                textField.text = String(format: "%@%@", textField.text!, symbol)
                symbol = self.symbolFor(textField.text?.count)!
                self.add(symbol, in: textField)
            }
            if range.location == self.numberCount + self.symbols.count + self.prefix.count {
                self.shake(textField:textField)
            }
            if range.length == 0 {
                return false
            }
        }
        
        // TextField is Empty
        if textField.text?.count == self.prefix.count {
            if self.numberCount + self.symbols.count > (textField.text?.count ?? 0) {
                if var symbol = self.symbolFor(self.prefix.count) {
                    if textField.text?.count != 0 {
                        textField.text = String(format:"%@%@", textField.text! ,symbol)
                    } else {
                        textField.text = String(format:"%@" , symbol)
                    }
                    symbol = self.symbolFor(textField.text?.count)!
                    self.add(symbol, in: textField)
                    
                    textField.text = String(format:"%@%@", textField.text!, string)
                    symbol = self.symbolFor(textField.text?.count)!
                    self.add(symbol, in: textField)
                } else {
                    textField.text = String(format:"%@", string)
                    let symbol = self.symbolFor(textField.text?.count)!
                    self.add(symbol, in: textField)
                }
            }
            return false
        }
        
        // When added symbol in textField
        if textField.text?.count == range.location {
            if self.numberCount + self.symbols.count + self.prefix.count > (textField.text?.count ?? 0) {
                if var symbol = self.symbolFor(textField.text?.count) {
                    textField.text = String(format:"%@%@", textField.text!, symbol)
                    symbol = self.symbolFor(textField.text?.count)!
                    self.add(symbol, in: textField)
                }
                if !self.isDelete && (textField.text?.count ?? 0) > 0 {
                    textField.text = String(format:"%@%@", textField.text!, string)
                }
                if var symbol = self.symbolFor(textField.text?.count) {
                    textField.text = String(format:"%@%@", textField.text!, symbol)
                    symbol = self.symbolFor(textField.text?.count)!
                    self.add(symbol, in: textField)
                }
                
                if self.isDelete {
                    textField.text = String(format:"%@%@", textField.text!, string)
                }
                return false
            }
            return false
        }
        
        // Delete text from textField
        if range.length > 0 {
            if range.location >= self.prefix.count {
                if textField.text?.count == range.location {
                    textField.text = textField.text?.substring(to: (textField.text?.endIndex)!)
                } else {
                    textField.text = textField.text?.substring(to: String.Index.init(encodedOffset: range.location))
                }
                if var symbol = self.symbolFor((textField.text?.count)! - 1) {
                    while(true) {
                        textField.text = textField.text!.substring(to: (textField.text?.endIndex)!)
                        if let _symbol = self.symbolFor((textField.text?.count)! - 1) {
                            symbol = _symbol
                            break
                        }
                    }
                }
            } else {
                textField.text = textField.text!.substring(to: String.Index.init(encodedOffset: self.prefix.count))
            }
            
            return false
        }
        return true
    }
    
    /// Set Contry code Prefix
    func setCountryName(_ name: String, textField: UITextField) {
        guard let code = CountryCode.countryCodeWith(name: name) else {
            print("We dosn't finde %@ .HB Default prefix is United States", name)
            self.prefix = CountryCode.countryCodeWith(name: "United States")!
            textField.text = String(format: "%@", self.prefix)
            return
        }
        self.prefix = code
        textField.text = String(format: "%@", code)
    }
    
    /// You can check your formatting is valid or not.
    func numberIsValidPhoneText(_ phoneText: String) -> Bool {
        let text = self.getNumber(phoneText)
        let length = text.count ?? 0
        if length == self.numberCount && self.isNumber(text) {
            return true
        } else {
            return false
        }
    }
    
    // MARK: Private API
    
    private var symbols: Dictionary<String, Any> = [String: Any]()
    private var numberCount: Int = 0
    private var isDelete: Bool = false
    private var prefix: String = ""
    
    private func isNumber(_ number: String) -> Bool {
        let digits = CharacterSet(charactersIn: "0123456789").inverted
        return NSString(string: number).rangeOfCharacter(from: digits).location == NSNotFound
    }
    
    private func shake(textField: UITextField) {
        if self.isShake {
            let shake = CABasicAnimation(keyPath:"position")
            shake.duration = CFTimeInterval(self.shakeDuration)
            shake.repeatCount = Float(self.shakeRepeatCount)
            shake.autoreverses = true
            shake.fromValue = NSValue(cgPoint: CGPoint(x: textField.center.x - self.shakeSize, y: textField.center.y))
            shake.toValue = NSValue(cgPoint: CGPoint(x: textField.center.x + self.shakeSize, y: textField.center.y))
            textField.layer.add(shake, forKey:"position")
        }
    }
    
    private func getNumber(_ mobileNumber: String?) -> String {
        let len = mobileNumber?.count ?? 0
        var buffer = [unichar]()
        NSString(string: mobileNumber!).getCharacters(&buffer, range:NSMakeRange(0, len))
        var number = ""
        for i in self.prefix.count..<len {
            let character = String(format:"%C", buffer[i])
            if self.isNumber(character) {
                if number.count != 0 {
                    number = String(format:"%@%@", number, character)
                } else {
                    number = String(format:"%@", character)
                }
            }
        }
        return number;
    }
    
    private func symbolFor(_ index: Int?) -> String? {
        let _index = (index ?? 0) - self.prefix.count
        for key in self.symbols.keys {
            if key == String(_index) {
                return self.symbols[key] as? String;
            }
        }
        return nil
    }
    
    private func add(_ symbol: String, in textField: UITextField) {
        var _symbol = symbol
        while _symbol.count > 0 {
            textField.text = String(format:"%@%@", textField.text!, symbol)
            _symbol = self.symbolFor(textField.text?.count)!
        }
    }

}
